File: kotlin-node/karakum/src/jsMain/resources/plugins/convertCompareResultType.js
import ts from "typescript";
import * as karakum from "../karakum.mjs";

export default function (node, context, render) {
    if (
        ts.isUnionTypeNode(node)
        && node.types.length === 3

        // -1
        && node.types[0]
        && ts.isLiteralTypeNode(node.types[0])
        && ts.isPrefixUnaryExpression(node.types[0].literal)
        && node.types[0].literal.operator === ts.SyntaxKind.MinusToken
        && ts.isNumericLiteral(node.types[0].literal.operand)
        && node.types[0].literal.operand.text === "1"

        // 0
        && node.types[1]
        && ts.isLiteralTypeNode(node.types[1])
        && ts.isNumericLiteral(node.types[1].literal)
        && node.types[1].literal.text === "0"

        // 1
        && node.types[2]
        && ts.isLiteralTypeNode(node.types[2])
        && ts.isNumericLiteral(node.types[2].literal)
        && node.types[2].literal.text === "1"
    ) {
        const typeScriptService = context.lookupService(karakum.typeScriptServiceKey.get())

        return `Int /* ${typeScriptService?.printNode(node)} */`
    }


    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/plugins/convertCompareResultType.kt
package node.karakum.plugins

import arrow.core.raise.nullable
import io.github.sgrishchenko.karakum.extension.createPlugin
import io.github.sgrishchenko.karakum.extension.plugins.TypeScriptService
import io.github.sgrishchenko.karakum.extension.plugins.typeScriptServiceKey
import typescript.SyntaxKind
import typescript.asArray
import typescript.isLiteralTypeNode
import typescript.isNumericLiteral
import typescript.isPrefixUnaryExpression
import typescript.isUnionTypeNode

val convertCompareResultType = createPlugin { node, context, _ ->
    nullable {
        ensure(isUnionTypeNode(node))

        val types = node.types.asArray()

        ensure(types.size == 3)

        // -1
        val minusOne = ensureNotNull(types[0])
        ensure(isLiteralTypeNode(minusOne))
        val minusOneLiteral = minusOne.literal
        ensure(isPrefixUnaryExpression(minusOneLiteral))
        ensure(minusOneLiteral.operator == SyntaxKind.MinusToken)
        val minusOneLiteralOperand = minusOneLiteral.operand
        ensure(isNumericLiteral(minusOneLiteralOperand))
        ensure(minusOneLiteralOperand.text == "1")

        // 0
        val zero = ensureNotNull(types[1])
        ensure(isLiteralTypeNode(zero))
        val zeroLiteral = zero.literal
        ensure(isNumericLiteral(zeroLiteral))
        ensure(zeroLiteral.text == "0")

        // 1
        val one = ensureNotNull(types[2])
        ensure(isLiteralTypeNode(one))
        val oneLiteral = one.literal
        ensure(isNumericLiteral(oneLiteral))
        ensure(oneLiteral.text == "1")

        val typeScriptService = ensureNotNull(context.lookupService(typeScriptServiceKey))

        "Int /* ${typeScriptService.printNode(node)} */"
    }
}
